### DISCUSSION TITLE
Implement rate limiter as a built-in middleware in Starlette

### DISCUSSION THREAD

**Discussion by magiskboy** [encode/starlette#2184] [Ideas] [answered]
I think we should implement a Rate Limiter Middleware class, which can be customized easily if need.

**Comment by zanieb** [comment]
Hi! Thanks for starting a discussion. Can you explain a bit more about what kind of rate limiting you're looking for, what kind of customization you'd want, and why it belongs in Starlette itself instead of a separate library?

**Comment by Kludex** [comment, marked as answer]
You can use [asgi-ratelimit](https://github.com/abersheeran/asgi-ratelimit).

Also, since latest releases, we support `middleware` parameter in `Route(middleware=)`, so it's easy to be as granular as you want.
